src/components/ui/sidebar.tsx

```diff
@@ -25,90 +25,110 @@ import {
+  expand: () => void
+  collapse: () => void
 }
 
 const SidebarContext = React.createContext<SidebarContextProps | null>(null)
 
-function useSidebar() {
-  const context = React.useContext(SidebarContext)
+// Enhanced context assertion function
+function assertSidebarContext(
+  context: SidebarContextProps | null
+): asserts context is SidebarContextProps {
   if (!context) {
-    throw new Error("useSidebar must be used within a SidebarProvider.")
+    throw new Error('useSidebar must be used within SidebarProvider')
   }
+}
 
+function useSidebar(): SidebarContextProps {
+  const context = React.useContext(SidebarContext)
+  assertSidebarContext(context)
   return context
 }
 
 const SidebarProvider = React.forwardRef<
   HTMLDivElement,
-  React.ComponentProps<"div"> & {
-    defaultOpen?: boolean
-    open?: boolean
-    onOpenChange?: (open: boolean) => void
-  }
+  SidebarProviderProps
 >(
   (
     {
-      defaultOpen = true,
-      open: openProp,
-      onOpenChange: setOpenProp,
+      defaultState,
+      state: stateProp,
+      onStateChange,
+      persistIntent = true,
+      responsiveBreakpoints,
       className,
       style,
       children,
       ...props
     },
     ref
   ) => {
-    const isMobile = useIsMobile()
-    const [openMobile, setOpenMobile] = React.useState(false)
-
-    // This is the internal state of the sidebar.
-    // We use openProp and setOpenProp for control from outside the component.
-    const [_open, _setOpen] = React.useState(defaultOpen)
-    const open = openProp ?? _open
-    const setOpen = React.useCallback(
-      (value: boolean | ((value: boolean) => boolean)) => {
-        const openState = typeof value === "function" ? value(open) : value
-        if (setOpenProp) {
-          setOpenProp(openState)
-        } else {
-          _setOpen(openState)
-        }
+    // Use our enhanced navigation drawer hook
+    const navigationDrawer = useNavigationDrawer()
+
+    // Convert enum state to boolean for legacy compatibility
+    const open = navigationDrawer.state === NavigationDrawerState.Expanded
+    const setOpen = React.useCallback((value: boolean | ((value: boolean) => boolean)) => {
+      const openState = typeof value === "function" ? value(open) : value
+      const newState = openState ? NavigationDrawerState.Expanded : NavigationDrawerState.Collapsed
+      navigationDrawer.setUserIntent(newState, StateChangeSource.UserAction)
+      onStateChange?.(newState, StateChangeSource.UserAction)
+    }, [open, navigationDrawer, onStateChange])
+
+    // Enhanced toggle that uses our new system with stable references
+    const toggleSidebar = React.useCallback(() => {
+      return navigationDrawer.isMobile
+        ? navigationDrawer.setOpenMobile(!navigationDrawer.openMobile)
+        : navigationDrawer.toggle(StateChangeSource.UserAction)
+    }, [navigationDrawer.isMobile, navigationDrawer.setOpenMobile, navigationDrawer.toggle, navigationDrawer.openMobile])
```

Comment:
Sections like these could be extracted into smaller named functions that tell the story of what's happening. When I read through it very carefully I get a sense of what's going on, but my mind finds smaller functions that describe what they're accomplishing more readable.

Reply:
A great suggestion. Fixed in commit 076b803d4cd84214a2c3a85e8b40079838502f51